### PR TITLE
[DAGCombiner] Preserve nneg flag from inner zext when we combine (z/s/aext (zext X))

### PR DESCRIPTION
**Pull request by topperc**

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -13710,8 +13710,12 @@ SDValue DAGCombiner::visitZERO_EXTEND(SDNode *N) {
 
   // fold (zext (zext x)) -> (zext x)
   // fold (zext (aext x)) -> (zext x)
-  if (N0.getOpcode() == ISD::ZERO_EXTEND || N0.getOpcode() == ISD::ANY_EXTEND)
-    return DAG.getNode(ISD::ZERO_EXTEND, DL, VT, N0.getOperand(0));
+  if (N0.getOpcode() == ISD::ZERO_EXTEND || N0.getOpcode() == ISD::ANY_EXTEND) {
+    SDNodeFlags Flags;
+    if (N0.getOpcode() == ISD::ZERO_EXTEND)
```
Comment (goldsteinn) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:13715]: Doesn't `zext` always imply nneg unless sizeof(dstty) <= sizeof(srcty) in which case its a nop?

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:13715]: nneg is about the sign bit of the input to the zero extend, not the output. It indicates that the zext could be a replaced with a sext.

**Review by nikic — COMMENTED** [review]
The changes LGTM, but is it possible to test it (or at least some parts of it)?

**Comment by topperc** [comment]
> The changes LGTM, but is it possible to test it (or at least some parts of it)?

Patch now contains changes to one of the tests from #82227. That tests the code in getNode(ISD::ANY_EXTEND

**Comment by topperc** [comment]
> > The changes LGTM, but is it possible to test it (or at least some parts of it)?
> 
> Patch now contains changes to one of the tests from #82227. That tests the code in getNode(ISD::ANY_EXTEND

Added two more tests to covert getNode for ISD::SIGN_EXTEND and ISD::ZERO_EXTEND.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**